**Kajetan Walawski - L2**

**Łączenie się z bazą danych za pomocą sqlAlchemy**

In [1]:
from sqlalchemy import create_engine

db_string = "postgresql://wbauer_adb:adb2020@pgsql-196447.vipserv.org:5432/wbauer_adb"

db = create_engine(db_string)

connection_sqlalchemy = db.connect()

**Łączenie się z bazą danych za pomocą psycopg2 i pandas**

In [2]:
import psycopg2 as pg
import pandas as pd

connection = pg.connect(host='pgsql-196447.vipserv.org', port=5432, dbname='wbauer_adb', user='wbauer_adb', password='adb2020')

**Ćwiczenie 1. Ile kategorii filmów mamy w wypożyczalni?**

In [3]:
cw1 = pd.read_sql('SELECT count(name) FROM category', con=connection)
odp1 = cw1.to_string(index=False, header=False)
print('Mamy', odp1, 'kategorii filmów w wypożyczalni.')

Mamy 16 kategorii filmów w wypożyczalni.


**Ćwiczenie 2. Wyświetl listę kategorii w kolejności alfabetycznej.**

In [4]:
cw2 = pd.read_sql('SELECT name FROM category ORDER BY name ASC',con=connection)
print(cw2.to_string(index=False, header=False))

     Action
  Animation
   Children
   Classics
     Comedy
Documentary
      Drama
     Family
    Foreign
      Games
     Horror
      Music
        New
     Sci-Fi
     Sports
     Travel


**Ćwiczenie 3. Znajdź najstarszy i najmłodszy film do wypożyczenia.**

In [5]:
cw3_old = pd.read_sql('SELECT title, release_year FROM film ORDER BY release_year DESC limit 1',con=connection)
cw3_new = pd.read_sql('SELECT title, release_year FROM film ORDER BY release_year ASC limit 1',con=connection)
print(cw3_old)
print("\n")
print(cw3_new)

             title  release_year
0  Chamber Italian          2006


             title  release_year
0  Chamber Italian          2006


Odpowiedź: Nie ma jednego najstarszego i jednego najmłodszego filmu, ponieważ wszystkie filmy do wypożyczenia są z 2006 roku.

**Ćwiczenie 4. Ile wypożyczeń odbyło się między 2005-07-01 a 2005-08-01?**

In [6]:
txt4 = """SELECT count(rental_date)
        FROM rental
        WHERE rental_date
        BETWEEN '2005-07-01 00:00:00'
        AND '2005-08-01 23:59:59'
        """
cw4 = pd.read_sql(txt4, con=connection)
odp4 = cw4.to_string(index=False, header=False)

print(cw4, '\n')
print('Między 2005-07-01 a 2005-08-01 włącznie odbyło się', odp4, 'wypożyczeń.')

   count
0   7380 

Między 2005-07-01 a 2005-08-01 włącznie odbyło się 7380 wypożyczeń.


**Ćwiczenie 5. Ile wypożyczeń odbyło się między 2010-01-01 a 2011-02-01?**

In [7]:
txt5 = """SELECT count(rental_date)
        FROM rental
        WHERE rental_date
        BETWEEN '2010-01-01 00:00:00' 
        AND '2011-02-01 23:59:59'
        """
cw5 = pd.read_sql(txt5, con=connection)
odp5 = cw5.to_string(index=False, header=False)

print(cw5, '\n')
print('Między 2010-01-01 a 2011-02-01 odbyło się', odp5, 'wypożyczeń.')

   count
0      0 

Między 2010-01-01 a 2011-02-01 odbyło się 0 wypożyczeń.


**Ćwiczenie 6. Znajdź największą płatność wypożyczenia.**

In [8]:
cw6 = pd.read_sql('SELECT * FROM payment ORDER BY amount DESC limit 1',con=connection)
cw6

payment_id  customer_id  staff_id  rental_id  amount  \
0       20403          362         1      14759   11.99   

                payment_date  
0 2007-03-21 21:57:24.996577

Największa płatność wypożyczenia wyniosła 11.99

**Ćwiczenie 7. Znajdź wszystkich klientów z Polski, Nigerii lub Bangladeszu.**

In [9]:
txt7 = """SELECT first_name, last_name, cntr.country
        FROM customer cust
        INNER JOIN address add ON add.address_id = cust.address_id
        INNER JOIN city ct ON ct.city_id = add.city_id
        INNER JOIN country cntr ON cntr.country_id = ct.country_id
        WHERE cntr.country IN ('Poland','Nigeria','Bangladesh')
        ORDER BY country
        """

cw7 = pd.read_sql(txt7, con=connection)
cw7

first_name  last_name     country
0    Michelle      Clark  Bangladesh
1     Stephen     Qualls  Bangladesh
2       Frank   Waggoner  Bangladesh
3     Marilyn       Ross     Nigeria
4       Elsie     Kelley     Nigeria
5      Gladys   Hamilton     Nigeria
6       Sonia    Gregory     Nigeria
7      Rodney    Moeller     Nigeria
8       Velma      Lucas     Nigeria
9       Carol     Garcia     Nigeria
10       Olga    Jimenez     Nigeria
11     Bertha   Ferguson     Nigeria
12     Tracey    Barrett     Nigeria
13         Jo     Fowler     Nigeria
14    Wallace      Slone     Nigeria
15  Constance       Reid     Nigeria
16      Brian      Wyman      Poland
17     Sidney   Burleson      Poland
18   Marjorie     Tucker      Poland
19    Russell    Brinson      Poland
20       Leah     Curtis      Poland
21      Ruben      Geary      Poland
22    Johnnie   Chisholm      Poland
23     Jimmie  Eggleston      Poland

**Ćwiczenie 8. Gdzie mieszkają członkowie personelu?**

In [10]:
txt8 = """SELECT first_name, last_name, add.address, ct.city, cntr.country
        FROM staff stf
        INNER JOIN address add ON add.address_id = stf.address_id
        INNER JOIN city ct ON ct.city_id = add.city_id
        INNER JOIN country cntr ON cntr.country_id = ct.country_id
        """

cw8 = pd.read_sql(txt8, con=connection)
cw8

first_name last_name               address        city    country
0       Mike   Hillyer     23 Workhaven Lane  Lethbridge     Canada
1        Jon  Stephens  1411 Lillydale Drive   Woodridge  Australia

**Ćwiczenie 9. Ilu pracowników mieszka w Argentynie lub Hiszpanii?**

In [11]:
txt9 = """SELECT count(last_name)
        FROM staff stf
        INNER JOIN address add ON add.address_id = stf.address_id
        INNER JOIN city ct ON ct.city_id = add.city_id
        INNER JOIN country cntr ON cntr.country_id = ct.country_id
        WHERE cntr.country IN ('Argentina','Spain')
        """

cw9 = pd.read_sql(txt9, con=connection)
print(cw9, '\n')
odp9 = cw9.to_string(index=False, header=False)
print('W Argentynie lub Hiszpanii mieszka', odp9, 'pracowników.')

   count
0      0 

W Argentynie lub Hiszpanii mieszka 0 pracowników.


**Ćwiczenie 10. Jakie kategorie filmów zostały wypożyczone przez klientów?**

In [12]:
txt10 = """SELECT distinct name
        FROM category cat
        INNER JOIN film_category f_cat ON cat.category_id = f_cat.category_id
        INNER JOIN film f ON f.film_id = f_cat.film_id
        INNER JOIN inventory inv ON inv.film_id = f.film_id
        INNER JOIN rental rent ON inv.inventory_id = rent.inventory_id
        """
cw10 = pd.read_sql(txt10, con=connection)
cw10

name
0        Sports
1      Classics
2           New
3        Family
4        Comedy
5     Animation
6        Travel
7         Music
8         Drama
9        Horror
10       Sci-Fi
11        Games
12  Documentary
13      Foreign
14       Action
15     Children

**Ćwiczenie 11. Znajdź wszystkie kategorie filmów wypożyczonych w Ameryce.**

In [13]:
txt11 = """SELECT distinct name
        FROM category cat
        INNER JOIN film_category f_cat ON f_cat.category_id = cat.category_id
        INNER JOIN film f ON f.film_id = f_cat.film_id
        INNER JOIN inventory inv ON inv.film_id = f.film_id
        INNER JOIN rental rent ON rent.inventory_id = inv.inventory_id
        INNER JOIN customer cust ON cust.customer_id = rent.customer_id
        INNER JOIN address add ON add.address_id = cust.address_id
        INNER JOIN city ct ON ct.city_id = add.city_id
        INNER JOIN country cntr ON cntr.country_id = ct.country_id
        WHERE cntr.country IN ('United States')
        """
cw11 = pd.read_sql(txt11, con=connection)
cw11

name
0        Action
1     Animation
2      Children
3      Classics
4        Comedy
5   Documentary
6         Drama
7        Family
8       Foreign
9         Games
10       Horror
11        Music
12          New
13       Sci-Fi
14       Sports
15       Travel

**Ćwiczenie 12. Znajdź wszystkie tytuły filmów, w których grał: Olympia Pfeiffer lub Julia Zellweger lub Ellen Presley.**

In [14]:
txt12 = """SELECT title, act.first_name, act.last_name
        FROM film f
        INNER JOIN film_actor f_act ON f.film_id = f_act.film_id
        INNER JOIN actor act ON act.actor_id = f_act.actor_id
        WHERE (act.first_name = 'Olympia' AND act.last_name = 'Pfeiffer')
        OR (act.first_name = 'Julia' AND act.last_name = 'Zellweger')
        OR (act.first_name = 'Ellen' AND act.last_name = 'Presley')
        """
pd.set_option('display.max_rows', None)
cw12 = pd.read_sql(txt12, con=connection)
cw12

title first_name  last_name
0          Bilko Anonymous      Ellen    Presley
1        Caribbean Liberty      Ellen    Presley
2         Casper Dragonfly      Ellen    Presley
3         Empire Malkovich      Ellen    Presley
4            Floats Garden      Ellen    Presley
5         Frogmen Breaking      Ellen    Presley
6           Homeward Cider      Ellen    Presley
7              Hyde Doctor      Ellen    Presley
8           Image Princess      Ellen    Presley
9            Jacket Frisco      Ellen    Presley
10    Microcosmos Paradise      Ellen    Presley
11            Network Peak      Ellen    Presley
12              Oscar Gold      Ellen    Presley
13          Pickup Driving      Ellen    Presley
14         Pinocchio Simon      Ellen    Presley
15            Private Drop      Ellen    Presley
16          Roots Remember      Ellen    Presley
17           Scarface Bang      Ellen    Presley
18         Secretary Rouge      Ellen    Presley
19                Spy Mile      Ellen    Presley
20    Streetcar Intentions      Ellen    Presley
21            Tadpole Park      Ellen    Presley
22        Treasure Command      Ellen    Presley
23               Turn Star      Ellen    Presley
24            Women Dorado      Ellen    Presley
25             Badman Dawn    Olympia   Pfeiffer
26             Chitty Lock    Olympia   Pfeiffer
27      Color Philadelphia    Olympia   Pfeiffer
28       Contact Anonymous    Olympia   Pfeiffer
29            Deep Crusade    Olympia   Pfeiffer
30        Effect Gladiator    Olympia   Pfeiffer
31          Express Lonely    Olympia   Pfeiffer
32       Firehouse Vietnam    Olympia   Pfeiffer
33        Fugitive Maguire    Olympia   Pfeiffer
34           Hanky October    Olympia   Pfeiffer
35            Ice Crossing    Olympia   Pfeiffer
36         Idols Snatchers    Olympia   Pfeiffer
37  Intolerable Intentions    Olympia   Pfeiffer
38      Magnolia Forrester    Olympia   Pfeiffer
39              Mars Roman    Olympia   Pfeiffer
40               Maude Mod    Olympia   Pfeiffer
41        Murder Antitrust    Olympia   Pfeiffer
42            None Spiking    Olympia   Pfeiffer
43             Others Soup    Olympia   Pfeiffer
44           Psycho Shrunk    Olympia   Pfeiffer
45             Santa Paris    Olympia   Pfeiffer
46             Sense Greek    Olympia   Pfeiffer
47         Storm Happiness    Olympia   Pfeiffer
48       Sweet Brotherhood    Olympia   Pfeiffer
49        Titanic Boondock    Olympia   Pfeiffer
50         Tourist Pelican    Olympia   Pfeiffer
51          Traffic Hobbit    Olympia   Pfeiffer
52              Wait Cider    Olympia   Pfeiffer
53    Breakfast Goldfinger      Julia  Zellweger
54        Cranes Reservoir      Julia  Zellweger
55             Dares Pluto      Julia  Zellweger
56        Detective Vision      Julia  Zellweger
57         Divorce Shining      Julia  Zellweger
58         Hollow Jeopardy      Julia  Zellweger
59         Jeopardy Encino      Julia  Zellweger
60        Lambs Cincinatti      Julia  Zellweger
61         Majestic Floats      Julia  Zellweger
62            Minds Truman      Julia  Zellweger
63            Open African      Julia  Zellweger
64            Outlaw Hanky      Julia  Zellweger
65         Panky Submarine      Julia  Zellweger
66        Rider Caddyshack      Julia  Zellweger
67               Won Dares      Julia  Zellweger
68           Wyoming Storm      Julia  Zellweger